controllers/ingress_controller.go

```diff
@@ -56,250 +61,190 @@ type IngressReconciler struct {
+				return nil
 			}
-		}
-
-		if !found {
-			apisToDelete.Items = append(apisToDelete.Items, x)
+			lg.Info("successful sync api defintion", "name", name, "op", op)
 		}
 	}
+	lg.Info("deleting orphan api's")
+	return r.deleteOrphanAPI(ctx, lg, ns, desired)
+}
 
-	// all the items to create
-	for _, x := range apisToCreateOrUpdate.Items {
-		create := true
-		for _, y := range apisToUpdate.Items {
-			if x.Name == y.Name {
-				create = false
-				break
-			}
-		}
-
-		if create {
-			apisToCreate.Items = append(apisToCreate.Items, x)
+func (r *IngressReconciler) deleteOrphanAPI(ctx context.Context, lg logr.Logger, ns string, desired *v1beta1.Ingress) error {
+	var keys []string
+	for _, rule := range desired.Spec.Rules {
+		for _, p := range rule.HTTP.Paths {
+			hash := shortHash(rule.Host + p.Path)
+			keys = append(keys, hash)
 		}
 	}
-
-	// create new endpoints first
-	for _, a := range apisToCreate.Items {
-		if err := r.Create(ctx, &a); err != nil {
-			log.Error(err, "unable to update api")
-		}
+	s := labels.NewSelector()
+	exists, err := labels.NewRequirement(apidefLabelKey, selection.Exists, []string{})
+	if err != nil {
+		return err
 	}
-
-	// update second
-	for _, a := range apisToUpdate.Items {
-		apiDefToUpdate := v1alpha1.ApiDefinition{}
-		err := r.Get(ctx, types.NamespacedName{Name: a.Name, Namespace: a.Namespace}, &apiDefToUpdate)
-		if err != nil {
-			log.Error(err, "unable to get api to update")
-			continue
-		}
-		apiDefToUpdate.Spec = a.Spec
-		if err := r.Update(ctx, &apiDefToUpdate); err != nil {
-			log.Error(err, "unable to update api")
-		}
+	s = s.Add(*exists)
+	notIn, err := labels.NewRequirement(apidefLabelKey, selection.NotIn, keys)
+	if err != nil {
+		return err
 	}
-
-	// delete last - just in-case something renamed
-	for _, a := range apisToDelete.Items {
-		if err := r.Delete(ctx, &a); err != nil {
-			log.Error(err, "unable to update api")
-		}
+	name, err := labels.NewRequirement(ingressLabelKey, selection.DoubleEquals, []string{desired.Name})
+	if err != nil {
+		return err
 	}
-
-	return ctrl.Result{}, nil
+	s.Add(*name)
+	s = s.Add(*notIn)
+	lg.Info("deleting orphan api definitions", "selector", s, "count", len(keys))
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		return r.DeleteAllOf(ctx, &v1alpha1.ApiDefinition{}, &client.DeleteAllOfOptions{
+			ListOptions: client.ListOptions{
+				LabelSelector: s,
+				Namespace:     ns,
+			},
+			DeleteOptions: client.DeleteOptions{},
+		})
+	})
 }
 
-func (r *IngressReconciler) buildAPIName(nameSpace, name, hostName, path string) string {
-	return fmt.Sprintf("%s-%s-%s", nameSpace, name, shortHash(hostName+path))
+func (r *IngressReconciler) buildAPIName(nameSpace, name, hash string) string {
+	return fmt.Sprintf("%s-%s-%s", nameSpace, name, hash)
 }
 
 func shortHash(txt string) string {
 	h := sha256.New()
 	h.Write([]byte(txt))
-	return fmt.Sprintf("%x", h.Sum(nil))[:9]
+	return fmt.Sprintf("%x", h.Sum(nil))[:4]
```

Comment:
Why change the API Name / Hash length? What is benefit?

Reply:
Nothing much, I just changed to what I originally suggested(first 4 bytes) . It looks much better.

Reply:
does this not increase the chance of collision?

Reply:
reverted back at 677be00